### PR TITLE
Fix Iterating over Nested Dictionary in For Loop AND Fix Capture Variable Support

### PR DESCRIPTION
**Pull request by danielcrabtree**
This PR fixes the issues discussed in #320.

### CODE REVIEW

**Comment by danielcrabtree** [comment]
This also now fixes #331

**Comment by microalps** [comment]
@daviburg This change is breaking, but it is the correct behavior IMHO. The test is changed to reference using item**.Value.**First

File: src/DotLiquid/Tags/Capture.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@ namespace DotLiquid.Tags
     /// </summary>
     public class Capture : DotLiquid.Block
     {
-        private static readonly Regex Syntax = R.C(@"(\w+)");
+        private static readonly Regex Syntax = R.B(R.Q(@"({0}+)"), Liquid.VariableSignature);
```
Comment (microalps) [inline src/DotLiquid/Tags/Capture.cs:25]: Variable signature includes dotted and bracketed notations. This should be a simple variable name imho. Other regex just include hyphen.

File: src/DotLiquid.Tests/Tags/StandardTagTests.cs [hunk 2]
```diff
@@ -81,6 +81,36 @@ public void TestForWithNestedDictionary()
                 }
             } };
 
+            Helper.AssertTemplateResult("JaneMike", "{% for item in People %}{{ item.Value.First }}{%endfor%}",
```
Comment (daviburg) [inline src/DotLiquid.Tests/Tags/StandardTagTests.cs:84]: Formatting should be cleaned-up.
If we move parameters to multiple lines, each parameter should start on a new line.

Comment (daviburg) [reply src/DotLiquid.Tests/Tags/StandardTagTests.cs:84]: ... and use named parameters as we have at least two strings.

File: src/DotLiquid.Tests/Tags/StandardTagTests.cs [hunk 3]
```diff
@@ -81,6 +81,36 @@ public void TestForWithNestedDictionary()
...
             } };
 
+            Helper.AssertTemplateResult("JaneMike", "{% for item in People %}{{ item.Value.First }}{%endfor%}",
+                Hash.FromDictionary(dictionary));
+        }
+
+        [Test]
+        public void TestForWithSingleLevelNestedDictionary()
+        {
+            var dictionary = new Dictionary<string, object> { {
```
Comment (daviburg) [inline src/DotLiquid.Tests/Tags/StandardTagTests.cs:91]: Fix the formatting here also. I assume that there are other use of nested bracketed collection initializers to refer to for consistency.

**Review by daviburg — CHANGES_REQUESTED** [review]

**Comment by daviburg** [comment]
> @daviburg This change is breaking, but it is the correct behavior IMHO. The test is changed to reference using item**.Value.**First

Sorry I don't see an existing test been changed in this PR. Could you help me understand the breaking change? What worked before that won't work after this?

**Comment by microalps** [comment]
> > @daviburg This change is breaking, but it is the correct behavior IMHO. The test is changed to reference using item**.Value.**First
> 
> Sorry I don't see an existing test been changed in this PR. Could you help me understand the breaking change? What worked before that won't work after this?

Yes its hard to see, but if you look carefully at the before and after, the old test for TestForWithNestedDictionary is on line 114 and new one is line 84. The old test remains, but now is applied to a different data set. The difference between these two lines is a ".Value". I missed it several times until I ran it locally.

**Comment by daviburg** [comment]
> > > @daviburg This change is breaking, but it is the correct behavior IMHO. The test is changed to reference using item**.Value.**First
> > 
> > 
> > Sorry I don't see an existing test been changed in this PR. Could you help me understand the breaking change? What worked before that won't work after this?
> 
> Yes its hard to see, but if you look carefully at the before and after, the old test for TestForWithNestedDictionary is on line 114 and new one is line 84. The old test remains, but now is applied to a different data set. The difference between these two lines is a ".Value". I missed it several times until I ran it locally.

I'm really dense today...
image

image

So before

```
        [Test]
        public void TestForWithNestedDictionary()
        {
            var dictionary = new Dictionary<string, object> { {
            "People", 
            new Dictionary<string, object> {
                    { "ID1", new Dictionary<string, object>{ { "First", "Jane" }, { "Last", "Green" } } },
                    { "ID2", new Dictionary<string, object>{ { "First", "Mike" }, { "Last", "Doe" } } }
                }
            } };

            Helper.AssertTemplateResult("JaneMike", "{% for item in People %}{{ item.First }}{%endfor%}",
                Hash.FromDictionary(dictionary));
        }
```

And after

```
        [Test]
        public void TestForWithNestedDictionary()
        {
            var dictionary = new Dictionary<string, object> { {
            "People", 
            new Dictionary<string, object> {
                    { "ID1", new Dictionary<string, object>{ { "First", "Jane" }, { "Last", "Green" } } },
                    { "ID2", new Dictionary<string, object>{ { "First", "Mike" }, { "Last", "Doe" } } }
                }
            } };

            Helper.AssertTemplateResult("JaneMike", "{% for item in People %}{{ item.Value.First }}{%endfor%}",
                Hash.FromDictionary(dictionary));
        }
```

Ah, ok, going from `{{ item.First }}` to `{{ item.Value.First }}`

Yeah we can't do that as `{{ item.First }}` was explicitly working before.

I didn't understand before that the 'star' characters were to highlight the keyword.

You are not using .First however in the new tests. Can as backward compatibility .First be a shortcut for .Value.First? I'm not really thinking this through, just tossing ideas.

**Comment by microalps** [comment]
From a performance perspective, I'm not a fan of it and just lets more people continue the same mistake without knowing. Maybe we can cut this code as part of a opt-in flag.

**Comment by microalps** [comment]
Closing this pull request as the bigger issue was resolved in another PR and the minor issue was incorrectly fixed as mentioned in code review.

**Pull request closed without merging**